File: app/src/main/java/com/shen/xi/android/tut/sinablog/QingPageDriver.java
package com.shen.xi.android.tut.sinablog;

import android.util.Log;

import com.google.inject.Inject;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;


public class QingPageDriver {
    private static final String TAG = QingPageDriver.class.getName();

    private List<String> mImageUrlList = new ArrayList<String>();

    @Inject
    public QingPageDriver() {
    }

    public boolean load(String url) {
        Document document = null;
        try {
            document = Jsoup.connect(url).get();
        } catch (IOException e) {
            return false;
        }

        parse(document);

        return true;
    }

    public boolean load(InputStream inputStream) {
        Document document = null;
        try {
            document = Jsoup.parse(inputStream, "utf-8", "qing.blog.sina.com.cn");
        } catch (IOException e) {
            Log.w(TAG, "cannot load page");

            return false;
        }

        parse(document);

        return true;
    }

    private void parse(Document document) {
        mImageUrlList.clear();
        if (document != null) {
            Elements mImageElement = document.select(".feedInfo .imgArea img");
            for (Element i : mImageElement) {
                if (i.hasAttr("real_src")) {
                    mImageUrlList.add(i.attr("real_src"));
                } else if (i.hasAttr("src")) {
                    mImageUrlList.add(i.attr("src"));
                } else {
                    Log.i(TAG, "img tag without src attribute");
                }
            }
        }
    }

    public List<String> getImageUrlList() {
        return mImageUrlList;
    }

}

File: app/src/main/scala/com/shen/xi/android/tut/sinablog/QingPageDriver.scala
package com.shen.xi.android.tut.sinablog

import java.io.{IOException, InputStream}
import java.util

import android.util.Log
import com.google.inject.Inject
import com.shen.xi.android.tut.util.ExtensionMethods
import org.jsoup.Jsoup
import org.jsoup.nodes.Document

import scala.collection.JavaConversions._

object QingPageDriver {
  val TAG = classOf[QingPageDriver].getName
}

class QingPageDriver @Inject()() {

  import ExtensionMethods._
  import com.shen.xi.android.tut.sinablog.QingPageDriver.TAG


  private val mImageUrlList: util.List[String] = new util.ArrayList[String]

  def getImageUrlList: util.List[String] = mImageUrlList

  def load(url: String): Boolean = try {
    Jsoup.connect(url).get() |> { l => if (l!=null) parse(l) }
    true
  } catch {
    case e: IOException => Log.w(TAG, s"cannot load page $url%s"); false
  }

  def load(inputStream: InputStream): Boolean = try {
    Jsoup.parse(inputStream, "utf-8", "qing.blog.sina.com.cn") |> { l => if (l != null) parse(l) }
    true
  } catch {
    case e: IOException => Log.w(TAG, "cannot parse page"); false
  }

  private def parse(document: Document): Unit = {
    mImageUrlList.clear()

    val urls = document select ".feedInfo .imgArea img" map (e => {
      if (e.hasAttr("real_src")) {
        e.attr("real_src")
      } else if (e.hasAttr("src")) {
        e.attr("src")
      } else {
        Log.w(TAG, "img tag without src attribute")
        null
      }
    }) filterNot (_ == null)

    mImageUrlList.addAll(urls)
  }

}
